src/logging.h

```diff
@@ -167,4 +168,50 @@ static inline void LogPrint(const BCLog::LogFlags& category, const Args&... args
     }
 }
 
+class ILogFilter
+{
+public:
+    virtual ~ILogFilter(){};
+    virtual bool filter() = 0;
+};
+
+// Implementation that logs at most every millis_ milliseconds
+class TimeThrottledFilter : public ILogFilter
+{
+private:
+    std::atomic_uint64_t millis_{0};
+    std::atomic_uint64_t last_time_millis_{0};
+
+public:
+    TimeThrottledFilter() = delete;
+    explicit TimeThrottledFilter(uint64_t millis) : millis_(millis)
+    {
+    }
+
+    bool filter() override
+    {
+        int64_t current_time = GetTimeMillis();
+        // First call or at least millis_ ms to last call
+        if ((last_time_millis_ == 0) || ((current_time - last_time_millis_) > millis_)) {
+            last_time_millis_ = current_time;
+            return false;
+        }
+        return true;
+    }
+};
+
+template <typename... Args>
+static inline void LogPrintThrottled(const BCLog::LogFlags& category, ILogFilter& time_throttled_filter, const Args&... args)
+{
+    // Log everything directly if category is enabled..
+    if (LogAcceptCategory((category))) {
+        LogPrintf(args...);
+    } else { // .. and otherwise time throttle
+        if (!time_throttled_filter.filter()) {
+            LogPrintf(args...);
+        }
+    }
+}
```

Comment:
I forgot to mention here, Category should be enabled and not filtered to be shown, the problem is throttle will override disabled category so it should be
```
if (LogAcceptCategory(category) && !time_throttled_filter.filter()) {
    LogPrintf(args...);
}
```

Reply:
I think this is intended and it's not overriding. The idea above proposed was that if the category is enabled it should log everything (no throttle) for detailed monitoring or debugging.
If the category is not enabled, you get the log throttled (e.g. every 10 minutes), so the debug.log is not flooded but still you get some information sometimes (e.g. the "waiting init" in the staker).

Reply:
But if you don't enable the category it should not throttle at all, so that's why we need category enabled + not filtered.

Reply:
The implementation is intended like that. If the category is enabled you want the full output (= no throttling). If the category is not enabled, it should log throttled (e.g. only every 10 minutes), to still have some information, but don't get a flooded debug.log.

This is what @prasannavl mentioned above:
> * Use a time throttled logger method for staking log, that throttles psuedo actionable info, like waiting reindex, waiting init and
>   unlock wallet.
> * Pass all those logs to the time throttled log, and for the time throttled log behaviour, if staking log is enabled, always log. And if it's not, throttle these.

Reply:
Hi @bvbfan and @prasannavl , do you have any updates on this improvement?
If I rebase to master, can we merge it? Or do you have more comments on it?
Thanks!